pkg/controller/apiextensions/composite/composed/composed.go

```diff
@@ -0,0 +1,181 @@
+}
+
+// Observation is the result of composed reconciliation.
+type Observation struct {
+	Ref               corev1.ObjectReference
+	ConnectionDetails managed.ConnectionDetails
+	Ready             bool
+}
+
+// WithClientApplicator returns a ComposerOption that changes the ClientApplicator of
+// Composer.
+func WithClientApplicator(ca resource.ClientApplicator) ComposerOption {
+	return func(composer *Composer) {
+		composer.client = ca
+	}
+}
+
+// WithConnectionDetailFetcher returns a ComposerOption that changes the
+// ConnectionDetailsFetcher of Composer.
+func WithConnectionDetailFetcher(cdf ConnectionDetailsFetcher) ComposerOption {
+	return func(composer *Composer) {
+		composer.ConnectionDetailsFetcher = cdf
+	}
+}
+
+// WithOverlayApplicator returns a ComposerOption that changes the
+// OverlayApplicator of Composer.
+func WithOverlayApplicator(oa OverlayApplicator) ComposerOption {
+	return func(composer *Composer) {
+		composer.OverlayApplicator = oa
+	}
+}
+
+// WithConfigurator returns a ComposerOption that changes the Configurator of
+// Composer.
+func WithConfigurator(c Configurator) ComposerOption {
+	return func(composer *Composer) {
+		composer.Configurator = c
+	}
+}
+
+// ComposerOption configures the Composer object.
+type ComposerOption func(*Composer)
+
+// NewComposer returns a new Composer that composes infrastructure resources
+// in a Kubernetes API server.
+func NewComposer(kube client.Client, opts ...ComposerOption) *Composer {
+	c := &Composer{
+		client: resource.ClientApplicator{
+			Client:     kube,
+			Applicator: resource.NewAPIPatchingApplicator(kube),
+		},
+		Configurator:             &DefaultConfigurator{},
+		OverlayApplicator:        &DefaultOverlayApplicator{},
+		ConnectionDetailsFetcher: &APIConnectionDetailsFetcher{client: kube},
+	}
+
+	for _, f := range opts {
+		f(c)
+	}
+	return c
+}
+
+// An Composer composes infrastructure resources in a Kubernetes API server.
+type Composer struct {
+	client resource.ClientApplicator
+	Configurator
+	OverlayApplicator
+	ConnectionDetailsFetcher
+}
+
+// Compose the supplied Composed resource into the supplied Composite resource
+// using the supplied CompositeTemplate.
+func (r *Composer) Compose(ctx context.Context, cp resource.Composite, cd resource.Composed, t v1alpha1.ComposedTemplate) (Observation, error) {
+
+	// Composed may not have a name if it's not created yet.
+	if cd.GetName() != "" {
+		nn := types.NamespacedName{Namespace: cd.GetNamespace(), Name: cd.GetName()}
+		if err := r.client.Get(ctx, nn, cd); resource.IgnoreNotFound(err) != nil {
+			return Observation{}, errors.Wrap(err, errGetComposed)
+		}
+	}
+
+	if !meta.WasCreated(cd) {
+		// Configure is called only if Composed resource is not created yet
+		// because the fields that are not referred in the patches should be
+		// mutable by the user, i.e. dependency on the Composition should be
+		// limited to the field propagation.
+		if err := r.Configure(cp, cd, t); err != nil {
```

Comment:
@negz Are you sure that we should unmarshal from template in each reconcile? Most of the time, this will make the composed resource immutable since any changes to fields that are not propagated from composite will be overwritten. We took this approach in templating-controller but it results in some problems, for example, if the composition had an array of tags and a late-init process wants to add something, the whole array is overwritten to the old values that exist in the composition, so, you get two controllers keep changing the same fields. This would happen for the propagated fields as well but that's more discoverable and I'd expect that usually much small number of fields would be propagated anyway compared to number of fields in the template.

I think, in general, it's desirable that we have the least number of spec fields being overwritten so that the resources are kept interactable, either by user or by another controller. Continuously enforcing the template greatly increases that number.

Other point I mentioned in the meeting is that changing a field on the primitive managed resource will become harder as it will require a new `Composition` resource to be created and pointed at; just for one field. When you consider that the person who creates the `Composition`s in the cluster, i.e. infra ops, the users of the resource will need them to do the change and `Composition` creation permission is an important one since it essentially enables people to provision anything, so infra-ops will always bear that cost of manual operation for the users. In other case, they just give the permission to work on specific composed resource kind(s). There could even be an RBAC controller which scans the cluster and allows SAs in the specific namespace to own the composed resources that are bound to requirements in that namespace to automate this. But there isn't much you can do if you have to have a `Composition` creation permission.

Reply:
That's probably subjective but it's easier for me to know that there is one single source of truth in my composed resources: requirement in my namespace. If its `Composition` is another source of truth, then I have to make sure the field I want to change does not appear in the template in `Composition` resource.

Reply:
Let me think about this and get back to you. I feel that the `kind: Composition` being considered on every reconcile is conceptually simpler (it seems like _fewer_ sources of truth to me), but I can see your point regarding conflicts with late initialization.

Reply:
Just capturing our decision from 1:1 today - I'm okay with this being merged as is for now because don't have a strong decision either way. I do still think it should be considered on every reconcile, but I want to write out my justification in a separate issue to avoid blocking this PR.

Reply:
Here's that issue I promised: https://github.com/crossplane/crossplane/issues/1481.

Reply:
@muvaf After writing up #1481 and thinking on it for a few days I'm feeling like I would prefer to consider the composition on every reconcile for the time being. I'm hoping to merge https://github.com/crossplane/crossplane/pull/1163 this week, and follow up with a design PR proposing the behaviour outlined in #1481.